collector/lib/ConfigLoader.cpp

```diff
@@ -8,114 +14,520 @@ namespace collector {
 namespace {
 const PathEnvVar CONFIG_FILE("ROX_COLLECTOR_CONFIG_PATH", "/etc/stackrox/runtime_config.yaml");
 
-enum PathTags {
+enum PathTags : uint8_t {
   LOADER_PARENT_PATH = 1,
   LOADER_CONFIG_FILE,
   LOADER_CONFIG_REALPATH,
 };
-}  // namespace
+
+std::string NodeTypeToString(YAML::NodeType::value type) {
+  // Don't add the default case so linters can warn about a missing type
+  switch (type) {
+    case YAML::NodeType::Null:
+      return "Null";
+    case YAML::NodeType::Undefined:
+      return "Undefined";
+    case YAML::NodeType::Scalar:
+      return "Scalar";
+    case YAML::NodeType::Sequence:
+      return "Sequence";
+    case YAML::NodeType::Map:
+      return "Map";
+  }
+  return "";  // Unreachable
+}
+};  // namespace
 
 namespace stdf = std::filesystem;
 
-ConfigLoader::ConfigLoader(CollectorConfig& config)
-    : config_(config), file_(CONFIG_FILE.value()) {}
+ParserResult ParserYaml::Parse(google::protobuf::Message* msg) {
+  YAML::Node node;
+  try {
+    node = YAML::LoadFile(file_);
+  } catch (const YAML::BadFile& e) {
+    return {{WrapError(e)}};
+  } catch (const YAML::ParserException& e) {
+    return {{WrapError(e)}};
+  }
 
-void ConfigLoader::Start() {
-  thread_.Start([this] { WatchFile(); });
-  CLOG(INFO) << "Watching configuration file: " << file_;
+  return Parse(msg, node);
 }
 
-void ConfigLoader::Stop() {
-  thread_.Stop();
-  CLOG(INFO) << "No longer watching configuration file: " << file_;
+ParserResult ParserYaml::Parse(google::protobuf::Message* msg, const YAML::Node& node) {
+  using namespace google::protobuf;
+
+  if (node.IsScalar() || node.IsNull()) {
+    return {{"Invalid configuration"}};
+  }
+
+  std::vector<ParserError> errors;
+
+  const Descriptor* descriptor = msg->GetDescriptor();
+  for (int i = 0; i < descriptor->field_count(); i++) {
+    const FieldDescriptor* field = descriptor->field(i);
+
+    auto err = Parse(msg, node, field);
+    if (err) {
+      errors.insert(errors.end(), err->begin(), err->end());
+    }
+  }
+
+  auto res = FindUnkownFields(*msg, node);
```

Comment:
Do we want to treat unknown fields in the same way as other errors. What if a user reverts to an older version, but keeps the same ConfigMap? If the new version introduced new fields, this would now cause an error.

Reply:
It might be better just to warn about unknown fields.

Reply:
This would require `ParserYaml` to have a notion of warning vs. error, it would require it to return something like a `std::pair<ParserError, ParserError>` or implement a `ParserError` with an error level. While this is achievable, it might be overkill and it will require some more changes and complexity to how `ParserYaml` itself is configured. Are you OK to keep things as are for now and we can tackle it in a follow up?

Side note, I have implemented a method to make errors ignorable in config-much itself, but I would prefer to keep us from escalating the scope of this PR even further.

- https://github.com/Molter73/config-much/commit/80beeb90391a1d7658f9ad16eae6a65b3130c6e2

Reply:
I think it could be done in a follow up PR, but I have a strong desire for that to be done before 4.8. Can you write the ticket? Thanks.

Reply:
I ended up adding the code to make it so unknown fields are ignored.